libsolidity/analysis/TypeChecker.cpp

```diff
@@ -599,8 +599,15 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 	if (auto referenceType = dynamic_cast<ReferenceType const*>(varType))
 	{
 		auto result = referenceType->validForLocation(referenceType->location());
-		if (result && (_variable.isConstructorParameter() || _variable.isPublicCallableParameter()))
-			result = referenceType->validForLocation(DataLocation::CallData);
+		if (result)
+		{
+			bool callDataCheckRequired =
+				(_variable.isLibraryFunctionParameter() && referenceType->location() == DataLocation::Storage) ? false :
+				(_variable.isConstructorParameter() || _variable.isPublicCallableParameter()) ? true :
+				false;
```

Comment:
I think it's better to refactor this nested ternary operator into ifs

Reply:
Also, the last `true: false;` is redundant.

Reply:
How about this?
```c++
bool isLibraryStorageParameter = (_variable.isLibraryFunctionParameter() && referenceType->location() == DataLocation::Storage);
if ((_variable.isConstructorParameter() || _variable.isPublicCallableParameter()) && !isLibraryStorageParameter)
```
or
```c++
bool libraryParameter = _variable.isLibraryFunctionParameter();
bool storageParameter = referenceType->location() == DataLocation::Storage;
bool constructorParameter = _variable.isConstructorParameter();
bool publicCallableParameter = _variable.isPublicCallableParameter();

if ((constructorParameter || publicCallableParameter) && !(libraryParameter && storageParameter))
```

Reply:
> I think it's better to refactor this nested ternary operator into ifs

I started with `if` and refactored to `?:` :). I agree of course that the expression can be simplified. I tried to make things easier to understand, but still reasonably short. OK, let's use @cameel's proposal,

```
bool isLibraryStorageParameter = (_variable.isLibraryFunctionParameter() && referenceType->location() == DataLocation::Storage);
bool callDataCheckRequired = ((_variable.isConstructorParameter() || _variable.isPublicCallableParameter()) && !isLibraryStorageParameter);
```

What we do now might be not that important, because #9746 reworks `callDataCheckRequired` anyway.